src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp

```diff
@@ -0,0 +1,458 @@
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+void PostgreSQLReplicationHandler::shutdown()
+{
+    stop_synchronization.store(true);
+    consumer_task->deactivate();
+}
+
+
+void PostgreSQLReplicationHandler::startSynchronization()
+{
+    createPublicationIfNeeded(connection->conn());
+
+    auto replication_connection = std::make_shared<PostgreSQLConnection>(fmt::format("{} replication=database", connection->conn_str()));
+    replication_connection->conn()->set_variable("default_transaction_isolation", "'repeatable read'");
+    auto ntx = std::make_shared<pqxx::nontransaction>(*replication_connection->conn());
+
+    std::string snapshot_name, start_lsn;
+
+    auto initial_sync = [&]()
+    {
+        createReplicationSlot(ntx, start_lsn, snapshot_name);
+        loadFromSnapshot(snapshot_name, storages);
+    };
+
+    /// Replication slot should be deleted with drop table only and created only once, reused after detach.
+    if (!isReplicationSlotExist(ntx, replication_slot))
+    {
+        initial_sync();
+    }
+    else if (!Poco::File(metadata_path).exists() || new_publication_created)
+    {
+        /// In case of some failure, the following cases are possible (since publication and replication slot are reused):
+        /// 1. If replication slot exists and metadata file (where last synced version is written) does not exist, it is not ok.
+        /// 2. If created a new publication and replication slot existed before it was created, it is not ok.
+        dropReplicationSlot(ntx);
+        initial_sync();
+    }
+    else
+    {
+        for (const auto & [table_name, storage] : storages)
+        {
+            try
+            {
+                nested_storages[table_name] = storage->getNested();
+                storage->setNestedLoaded();
+            }
+            catch (...)
+            {
+                tryLogCurrentException(__PRETTY_FUNCTION__);
+            }
+        }
+    }
+
+    ntx->commit();
+
+    consumer = std::make_shared<PostgreSQLReplicaConsumer>(
+            context,
+            connection,
+            replication_slot,
+            publication_name,
+            metadata_path,
+            start_lsn,
+            max_block_size,
+            allow_minimal_ddl,
+            is_postgresql_replica_database_engine,
+            nested_storages);
+
+    consumer_task->activateAndSchedule();
+
+    replication_connection->conn()->close();
+}
+
+
+NameSet PostgreSQLReplicationHandler::loadFromSnapshot(std::string & snapshot_name, Storages & sync_storages)
+{
+    NameSet success_tables;
+    for (const auto & storage_data : sync_storages)
+    {
+        try
+        {
+            auto tx = std::make_shared<pqxx::work>(*connection->conn());
+            const auto & table_name = storage_data.first;
+
+            /// Specific isolation level is required to read from snapshot.
+            tx->set_variable("transaction_isolation", "'repeatable read'");
```

Comment:
```suggestion
            /// Specific isolation level is required to read from snapshot.
            auto tx = std::make_shared<pqxx::transaction<isolation_level::repeatable_read, write_policy::read_only>>(*connection->conn());
            const auto & table_name = storage_data.first;
```

Reply:
I am not sure whether it is better, because it is passed to several methods in different places and therefore requires changing small `pqxx::work` to a long <pqxx::transaction<isolation_level::repeatable_read, write_policy::read_only> in several places, and it looks less readable and a little bit ugly.

Reply:
```
using NiceShortName = some_namespace::ugly_template<other_namespase::with_long_name, and_some<unreadable_parameters>>;
using NiceShortNamePtr = std::shared_ptr<NiceShortName>;
```